rpc/codec/server.go

```diff
@@ -79,6 +92,72 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// userFromRequest takes a request proto and attemps to access
+// request.RequestHeader.User.
+func (c *serverCodec) userFromRequest(request proto.Message) (string, error) {
+	// Extract Header.User from the request, excluding Ping requests.
+	// TODO: How much of this do we want to split into separate checks?
+	header := reflect.ValueOf(request).Elem().FieldByName("RequestHeader")
+	if !header.IsValid() {
+		return "", util.Errorf("missing RequestHeader in request: %s",
+			reflect.ValueOf(request))
+	}
+
+	hdr, ok := header.Interface().(cockroach_proto.RequestHeader)
+	if !ok {
+		return "", util.Errorf("bad RequestHeader in request: %s", header.Type())
+	}
+
+	if len(hdr.User) == 0 {
+		return "", util.Errorf("missing User in request header: %s", hdr)
+	}
+
+	return hdr.User, nil
+}
+
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// For requests with a RequestHeader, we compare the header.User against
+// the client certificate Subject.CommonName.
+// For other requests, we either allow all (test requests), or
+// require the "node user".
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
+	var err error
+	var requestedUser string
+
+	switch request := request.(type) {
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest:
+		// Ping and Gossip requests do not have a header: require the node user.
+		requestedUser = security.NodeUser
+	case *message.EchoRequest, *message.ArithRequest:
+		// These are for testing purposes only.
+		return nil
+	default:
```

Comment:
oh yeah, much better. I wasn't aware of proto.Request.
This simplifies everything.